### DISCUSSION TITLE
Suggestion: Sound Cards

### DISCUSSION THREAD

**Discussion by ohne-ende** [geerlingguy/raspberry-pi-pcie-devices#34] [Ideas] [answered]
I had an idea to build a music setup with the CM4 and because audio on Raspberry Pi is not as great as I wanted, I wondered if sound cards connected to the PCIe interface would work.

**Comment by P-Mouse** [comment]
I was send here by Jeff :)

I actually see some nice use cases for soundcards myself.
One big issue is to get surround-sound or mutli-channel output.
The last one can be used for like a active loudspeaker systems.
A lab-measuring system can also easily made this way

By default the I2S can be used, but this is only capable for 2-channel audio.
Technically the main processor is capable of two lanes, but one is actually not in use if I am not mistaken.
USB soundcard could also be used, but you pay quite a lot for getting multichannel 24bit/96kHz audio
(yes I know that not in every case this is noticeable, but it's about that people just like to use it)

This can get even more interesting when you use a PCIE to PCI adapter.
There are quite a lot of very good performing 2nd hand soundcards available. 

I don't have direct experience with patching and running my own kernels.
But my experience is that Linux is quite easy with soundscards in general.

**Comment by geerlingguy** [comment, marked as answer]
FYI there's at least one card confirmed to work with the Pi, the Creative SB1040. I've added a new section to the site, "Multimedia and Sound Cards" (https://pipci.jeffgeerling.com/#multimedia-av-and-sound-cards), and will add any other cards people introduce there.

**Reply by 6by9** [reply]
Avoid C-Media CMI8738 based cards for now. They're actually a PCIe to PCI bridge, and then a PCI audio chip. The PCI side requires I/O address space only, so doesn't work on ARM. 
The driver (under SND_CMIPCI) tries to configure the card via [```in``` and ```out``` I/O commands](https://github.com/raspberrypi/linux/blob/rpi-5.10.y/sound/pci/cmipci.c#L502). That driver is common for C-Media CMI8338, CMI8738, CMI8768 and CMI8770 chips.

C-Media 8786, 8787, 8788 are a different driver (SND_OXYGEN), but it also still retains references to ```in``` and ```out``` operations.

CM8826 and CM8828 appear to be Intel HDA compliant, so have a reasonable chance of working.

I've previously tried an M-Audio Audiophile 2496 (and Delta 410 IIRC) via PCIe to a PCI bridge, and again they fail due to trying to use I/O space on ARM.

**Reply by P-Mouse** [reply]
Thanks a lot for the update!

Looking forward to see some more cards!
If memory serves well, there are a lot of Creative cards that works like the SB1040
I wonder about the Asus Xonar DX cards, they are very easy in Linux in general.

**Reply by 6by9** [reply]
"Asus Xonar DG/DGX" is listed under the same driver as C-Media 8786/8787/8788 (SND_OXYGEN).

"Asus AV66/AV100/AV200 chips, i.e., Xonar D1, DX, D2, D2X, DS, DSX, Essence ST (Deluxe), and Essence STX (II)" are under SND_VIRTUOSO which shares large lumps of SND_OXYGEN (via the internal option SND_OXYGEN_LIB), so I wouldn't hold my breath on them.
```lspci -vvv``` on a machine with one of these cards in would give the information needed (may need sudo for all the info).
